File: app/src/main/java/com/example/peyo/gles20/Grid.java
package com.example.peyo.gles20;

import android.opengl.GLES20;
import android.opengl.Matrix;

import java.nio.FloatBuffer;

public class Grid extends GLES20 {

    static float coords[] = {
            -4,0,0, 4,0,0,  0,0,-4, 0,0,4};

    static float h = 0.5f;
    static float colors[] = {
            h,h,h,1, h,h,h,1, h,h,h,1, h,h,h,1};

    private FloatBuffer vertexBuffer;
    private FloatBuffer colorBuffer;

    private int mPositionHandle;
    private int mColorHandle;
    private int mMVPMatrixHandle;
    private float[] mMVPMatrix;

    public Grid(int position, int color, int mvp) {
        vertexBuffer = GLToolbox.loadBuffer(coords);
        colorBuffer = GLToolbox.loadBuffer(colors);

        mPositionHandle = position;
        mColorHandle = color;
        mMVPMatrixHandle = mvp;
    }

    public void draw(float[] mvpMatrix) {
        mMVPMatrix = mvpMatrix;

        glVertexAttribPointer(mPositionHandle, 3, GL_FLOAT, false,
                3 * 4, vertexBuffer);
        glVertexAttribPointer(mColorHandle, 4, GL_FLOAT, false,
                4 * 4, colorBuffer);
        glLineWidth(3);

        for (float f=-4; f<=4; f++) {
            translate(0, 0, f);
            glDrawArrays(GL_LINES, 0, 2);

            translate(f, 0, 0);
            glDrawArrays(GL_LINES, 2, 2);
        }
    }

    private void translate(float x, float y, float z) {
        float[] mvpMatrix = new float[16];
        Matrix.translateM(mvpMatrix, 0, mMVPMatrix, 0, x, y, z);
        GLES20.glUniformMatrix4fv(mMVPMatrixHandle, 1, false, mvpMatrix, 0);
    }
}

File: app/src/main/java/com/example/peyo/gles20/Grid.kt
package com.example.peyo.gles20

import android.opengl.GLES20
import android.opengl.Matrix
import java.nio.FloatBuffer


class Grid(val positionHandle: Int, val colorHandle: Int, val matrixHandle: Int) : GLES20() {
    val coords = floatArrayOf (-4f,0f,0f,  4f,0f,0f,  0f,0f,-4f,  0f,0f,4f )
    val h = 0.5f;
    val colors = floatArrayOf (h,h,h,1f,  h,h,h,1f,  h,h,h,1f,  h,h,h,1f)
    private val vertexCount: Int = coords.size / 3
    private val vertexBuffer: FloatBuffer = GLToolbox.loadBuffer(coords)
    private val colorBuffer: FloatBuffer = GLToolbox.loadBuffer(colors)

    fun draw(vpMatrix: FloatArray) {
        glVertexAttribPointer(positionHandle, 3, GL_FLOAT, false,
                3 * 4, vertexBuffer)
        glVertexAttribPointer(colorHandle, 4, GL_FLOAT, false,
                4 * 4, colorBuffer)
        glLineWidth(3f)

        for (i in -4 .. 4) {
            translate(vpMatrix, 0f, 0f, i.toFloat())
            glDrawArrays(GL_LINES, 0, 2)
            translate(vpMatrix, i.toFloat(),0f, 0f)
            glDrawArrays(GL_LINES, 2, 2)
        }
    }

    private fun translate(vpMatrix: FloatArray, x: Float, y: Float, z: Float) {
        val transMatrix = FloatArray(16)
        val vpMatrix2 = FloatArray(16)
        Matrix.setIdentityM(transMatrix, 0)
        Matrix.translateM(transMatrix, 0, x, y, z)
        Matrix.multiplyMM(vpMatrix2, 0, vpMatrix, 0, transMatrix, 0)
        glUniformMatrix4fv(matrixHandle, 1, false, vpMatrix2, 0)
    }
}
